lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts

```diff
@@ -61,47 +111,341 @@ export class UnifiedCacheManager {
+    private getCredentialsFilteredByCredentialType(
+        cacheCredentials: object,
+        homeAccountId?: string,
+        environment?: string,
+        credentialType?: string,
+        clientId?: string,
+        realm?: string,
+        target?: string
+    ): Object {
+        const matchingCredentials = {};
+        let matches: boolean = true;
+
+        Object.keys(cacheCredentials).forEach((key) => {
+            if (!!homeAccountId) {
+                matches = CacheHelper.matchHomeAccountId(
+                    key,
+                    homeAccountId
+                );
+            }
+
+            if (!!environment) {
+                matches = matches && CacheHelper.matchEnvironment(key, environment);
+            }
+
+            if (!!realm) {
+                matches = matches && CacheHelper.matchRealm(key, realm);
+            }
+
+            if (!!credentialType) {
+                matches = matches && CacheHelper.matchCredentialType(key, credentialType);
+            }
+
+            if (!!clientId) {
+                matches = matches && CacheHelper.matchClientId(key, clientId);
+            }
+
+            // idTokens do not have "target", target specific refreshTokens do exist for some types of authentication
+            if (!!target && CacheHelper.getCredentialType(key) != CredentialType.ID_TOKEN.toString().toLowerCase()) {
+                matches = matches && CacheHelper.matchTarget(key, target);
+            }
+
+            if (matches) {
+                matchingCredentials[key] = cacheCredentials[key];
+            }
+        });
+
+        return matchingCredentials;
     }
 
     /**
-     * append account to the in memory cache
+     * returns a boolean if the given account is removed
      * @param account
      */
-    addAccountEntity(account: AccountEntity): void {
-        const accKey = account.generateAccountKey();
-        if (!this.inMemoryCache.accounts[accKey]) {
-            this.inMemoryCache.accounts[accKey] = account;
+    removeAccount(account: AccountEntity): boolean {
+        const cache = this.getCacheInMemory();
+        const accountKey = account.generateAccountKey();
+
+        if (!!cache.accounts[accountKey]) {
+            delete cache.accounts[accountKey];
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * returns a boolean if the given account is removed
+     * @param account
+     */
+    removeAccountContext(account: AccountEntity): boolean {
+        const cache = this.getCacheInMemory();
+
+        const accountId = account.generateAccountId();
+
+        // TODO: Check how this should be done, do we just remove the account or also the associated credentials always?
+        Object.keys(cache.accessTokens).forEach((key) => {
+            if (
+                cache.accessTokens[key].generateAccountId() === accountId
+            )
+                this.removeCredential(cache.accessTokens[key]);
```

Comment:
Can you add curly brackets here (this will be enforced by linting soon).

Reply:
Done